src/cpp/shared/game/game.cpp

```diff
@@ -268,130 +267,124 @@ void game::ProcessOnMapHeroes() {
+          mapExtraHero->couldBeHasFaction = 0;
+        } else {
+          mightBeHeroAlreadyExists[mapExtraHero->heroID] = 1;
+          mapExtraHero->couldBeHasFaction = 1;
+        }
+
+        if (isJail) {
+          mapExtraHero->owner = -1;
+          faction = mapExtraHero->factionID;
+        } else {
+          mapExtraHero->owner = gcColorToPlayerPos[loc->objectIndex / 7];
+          faction = loc->objectIndex % 7; // Constant here (needs to be generalized)
+          if (faction == FACTION_MULTIPLE) { // Constant here (faction related)
+            faction = this->relatedToColorOfPlayerOrFaction[gcColorToSetupPos[gpGame->players[mapExtraHero->owner].color]];// Constant here (I think this is a faction-related structure)
           }
+        }
 
-          if (i == 1) {                       // 2nd iteration
-            if (isJail) {
-              faction = mapExtraHero->factionID;
-            } else {
-              faction = loc->objectIndex % 7; // Constant here (needs to be generalized)
-              if (faction == FACTION_MULTIPLE)// Constant here (faction related)
-                faction = this->relatedToColorOfPlayerOrFaction[gcColorToSetupPos[gpGame->players[mapExtraHero->owner].color]];// Constant here (most likely faction related)
-            }
-            if (mapExtraHero->couldBeHasFaction) {
-              this->heroes[mapExtraHero->heroID].factionID = faction;
-            } else {
-              randomHeroIdx = this->RandomScan((signed char *)mightBeHeroAlreadyExists, 9 * faction, 9, 1000, 0);// Constant here (the game might depend on the number of heroes as it relates to the number of factions)
-              if (randomHeroIdx == -1) { //  I think RandomScan is just a strange way of trying to return a random Idx that will correspond to a hero that satisfies a particular criterion, yet I think there is a better way of accomplishing this.
-                randomHeroIdx = this->RandomScan((signed char *)mightBeHeroAlreadyExists, 0, 54, 10000, 0);// Constant here (the game might depend on the number of heroes as it relates to the number of factions)
-                faction = randomHeroIdx / 9; // Constant here (relies on relation between number of factions and number of heroes)
-              }
-              mightBeHeroAlreadyExists[randomHeroIdx] = 1;
-              this->heroes[randomHeroIdx].factionID = faction;
-              if (mapExtraHero->field_11 && mapExtraHero->heroID >= 54)
-                this->heroes[randomHeroIdx].heroID = mapExtraHero->heroID;
-              mapExtraHero->heroID = randomHeroIdx;
-            }
+        if (mapExtraHero->couldBeHasFaction) {
+          this->heroes[mapExtraHero->heroID].factionID = faction;
+        } else {
+          randomHeroIdx = this->RandomScan((signed char *)mightBeHeroAlreadyExists, 9 * faction, 9, 1000, 0);// Constant here (the game might depend on the number of heroes as it relates to the number of factions)
```

Comment:
I think both of these 9's are something like HEROES_PER_FACTION.

Reply:
That makes sense. I'm still not confident what is so significant about the values passed into RandomScan (I think it tries to find a random that isn't covered in mightBeHeroAlreadyExists) and I'm also not confident that this will work well with the new factions.

Reply:
The first argument is presumably an array of heroes that have already been used.

The second and third arguments give bounds on the hero IDs that it may use.

The fourth argument is the number of random tries, presumably.
The last argument is ???.

There are a number of "special" heroes that have the next 12 or so IDs. The Cyborg heroes would need to be after that.

Reply:
I find the use of this function to be rather strange. I have also noticed that the left bound can actually exceed the right bound in the first call to this function. Additionally I'm not sure what special heroes you are talking about and/or what you mean by "the next 12".

Reply:
There are certain heroes such as Brother Brax, Sister Carlawna, Archibald, Roland, and Solmyr which only appear in the campaigns, or when specified in the map editor. They never spawn randomly.

Reply:
Ah. It sounds like I'm going to need to learn more about the heroes to generalize this properly.